File: backend/internal/models/homepage.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// SectionType represents the type of homepage section
type SectionType string

const (
	SectionBannerCarousel SectionType = "banner_carousel"
	SectionCategories     SectionType = "categories"
	SectionDealOfDay      SectionType = "deal_of_day"
	SectionFlashSale      SectionType = "flash_sale"
	SectionBestSellers    SectionType = "best_sellers"
	SectionFeatured       SectionType = "featured"
	SectionNewArrivals    SectionType = "new_arrivals"
	SectionSpecialOffers  SectionType = "special_offers"
	SectionBrands         SectionType = "brands"
	SectionRecentlyViewed SectionType = "recently_viewed"
	SectionUpcomingEvents SectionType = "upcoming_events"
	SectionCustomBanner   SectionType = "custom_banner"
	SectionShowcase360    SectionType = "showcase_360"
	SectionBundleDeals    SectionType = "bundle_deals"
)

// HomepageSection represents a configurable section on the homepage
type HomepageSection struct {
	ID          primitive.ObjectID     `json:"id" bson:"_id,omitempty"`
	Type        SectionType            `json:"type" bson:"type"`
	Title       string                 `json:"title" bson:"title"`
	Subtitle    string                 `json:"subtitle" bson:"subtitle"`
	Priority    int                    `json:"priority" bson:"priority"` // Lower number = higher priority
	IsActive    bool                   `json:"isActive" bson:"isActive"`
	Config      map[string]interface{} `json:"config" bson:"config"` // Section-specific configuration
	StartDate   *time.Time             `json:"startDate" bson:"startDate"`
	EndDate     *time.Time             `json:"endDate" bson:"endDate"`
	CreatedBy   primitive.ObjectID     `json:"createdBy" bson:"createdBy"`
	CreatedAt   time.Time              `json:"createdAt" bson:"createdAt"`
	UpdatedAt   time.Time              `json:"updatedAt" bson:"updatedAt"`
}

// HomepageConfig represents the complete homepage configuration
type HomepageConfig struct {
	ID        primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Sections  []HomepageSection  `json:"sections" bson:"sections"`
	UpdatedBy primitive.ObjectID `json:"updatedBy" bson:"updatedBy"`
	UpdatedAt time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// SectionLayoutItem represents the order and visibility of a section
type SectionLayoutItem struct {
	SectionType SectionType `json:"sectionType" bson:"sectionType"`
	Order       int         `json:"order" bson:"order"`           // Display order (0 = first, 1 = second, etc.)
	IsVisible   bool        `json:"isVisible" bson:"isVisible"`   // Whether section is visible
	Title       string      `json:"title,omitempty" bson:"title,omitempty"` // Optional custom title override
}

// HomepageLayout represents the display order configuration for homepage sections
type HomepageLayout struct {
	ID        primitive.ObjectID  `json:"id" bson:"_id,omitempty"`
	Layout    []SectionLayoutItem `json:"layout" bson:"layout"`
	UpdatedBy primitive.ObjectID  `json:"updatedBy" bson:"updatedBy"`
	UpdatedAt time.Time           `json:"updatedAt" bson:"updatedAt"`
	CreatedAt time.Time           `json:"createdAt" bson:"createdAt"`
}

// DealOfDay represents a time-limited special deal
type DealOfDay struct {
	ID           primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	ProductID    primitive.ObjectID `json:"productId" bson:"productId"`
	OriginalPrice float64           `json:"originalPrice" bson:"originalPrice"`
	DealPrice    float64            `json:"dealPrice" bson:"dealPrice"`
	DiscountPercent int             `json:"discountPercent" bson:"discountPercent"`
	StartTime    time.Time          `json:"startTime" bson:"startTime"`
	EndTime      time.Time          `json:"endTime" bson:"endTime"`
	Stock        int                `json:"stock" bson:"stock"`
	SoldCount    int                `json:"soldCount" bson:"soldCount"`
	IsActive     bool               `json:"isActive" bson:"isActive"`
	CreatedAt    time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt    time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// FlashSale represents a time-limited sale event
type FlashSale struct {
	ID          primitive.ObjectID   `json:"id" bson:"_id,omitempty"`
	Title       string               `json:"title" bson:"title"`
	Description string               `json:"description" bson:"description"`
	BannerImage string               `json:"bannerImage" bson:"bannerImage"`
	ProductIDs  []primitive.ObjectID `json:"productIds" bson:"productIds"`
	StartTime   time.Time            `json:"startTime" bson:"startTime"`
	EndTime     time.Time            `json:"endTime" bson:"endTime"`
	Discount    int                  `json:"discount" bson:"discount"` // Percentage
	IsActive    bool                 `json:"isActive" bson:"isActive"`
	CreatedAt   time.Time            `json:"createdAt" bson:"createdAt"`
	UpdatedAt   time.Time            `json:"updatedAt" bson:"updatedAt"`
}

// Brand represents a jewelry brand
type Brand struct {
	ID          primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Name        string             `json:"name" bson:"name"`
	Logo        string             `json:"logo" bson:"logo"`
	Description string             `json:"description" bson:"description"`
	IsActive    bool               `json:"isActive" bson:"isActive"`
	Priority    int                `json:"priority" bson:"priority"`
	CreatedAt   time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt   time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// RecentlyViewed tracks products viewed by users
type RecentlyViewed struct {
	ID         primitive.ObjectID   `json:"id" bson:"_id,omitempty"`
	UserID     *primitive.ObjectID  `json:"userId" bson:"userId"`
	SessionID  *string              `json:"sessionId" bson:"sessionId"` // For guest users
	ProductIDs []primitive.ObjectID `json:"productIds" bson:"productIds"`
	UpdatedAt  time.Time            `json:"updatedAt" bson:"updatedAt"`
}

// Showcase360 represents an interactive 360° product showcase
type Showcase360 struct {
	ID          primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	ProductID   primitive.ObjectID `json:"productId" bson:"productId"`
	Title       string             `json:"title" bson:"title"`
	Description string             `json:"description" bson:"description"`
	Images360   []string           `json:"images360" bson:"images360"`     // Array of images for 360° rotation
	VideoURL    string             `json:"videoUrl" bson:"videoUrl"`       // Video loop of jewelry being worn
	ThumbnailURL string            `json:"thumbnailUrl" bson:"thumbnailUrl"` // Thumbnail for the showcase
	Priority    int                `json:"priority" bson:"priority"`       // Display order
	IsActive    bool               `json:"isActive" bson:"isActive"`
	StartTime   *time.Time         `json:"startTime" bson:"startTime"`
	EndTime     *time.Time         `json:"endTime" bson:"endTime"`
	CreatedAt   time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt   time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// BundleItem represents a product in a bundle
type BundleItem struct {
	ProductID primitive.ObjectID `json:"productId" bson:"productId"`
	Quantity  int                `json:"quantity" bson:"quantity"`
}

// BundleDeal represents a bundle of products sold together at a discount
type BundleDeal struct {
	ID              primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Title           string             `json:"title" bson:"title"`
	Description     string             `json:"description" bson:"description"`
	BannerImage     string             `json:"bannerImage" bson:"bannerImage"`
	Items           []BundleItem       `json:"items" bson:"items"`
	OriginalPrice   float64            `json:"originalPrice" bson:"originalPrice"`
	BundlePrice     float64            `json:"bundlePrice" bson:"bundlePrice"`
	DiscountPercent int                `json:"discountPercent" bson:"discountPercent"`
	Category        string             `json:"category" bson:"category"` // e.g., "Bridal", "Complete Look", "Gift Set"
	Priority        int                `json:"priority" bson:"priority"`
	IsActive        bool               `json:"isActive" bson:"isActive"`
	StartTime       *time.Time         `json:"startTime" bson:"startTime"`
	EndTime         *time.Time         `json:"endTime" bson:"endTime"`
	Stock           int                `json:"stock" bson:"stock"`
	SoldCount       int                `json:"soldCount" bson:"soldCount"`
	CreatedAt       time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt       time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// HomepageResponse represents the complete homepage data response
type HomepageResponse struct {
	Layout           []SectionLayoutItem `json:"layout,omitempty"`        // Section ordering configuration
	Sections         []HomepageSection   `json:"sections"`
	DealOfDay        *DealOfDay          `json:"dealOfDay,omitempty"`
	ActiveFlashSales []FlashSale         `json:"activeFlashSales,omitempty"`
	Brands           []Brand             `json:"brands,omitempty"`
	RecentlyViewed   []Product           `json:"recentlyViewed,omitempty"`
	Showcases360     []Showcase360       `json:"showcases360,omitempty"`
	BundleDeals      []BundleDeal        `json:"bundleDeals,omitempty"`
}

// Helper methods

// IsCurrentlyActive checks if a section is currently active based on dates
func (s *HomepageSection) IsCurrentlyActive() bool {
	if !s.IsActive {
		return false
	}

	now := time.Now()

	if s.StartDate != nil && now.Before(*s.StartDate) {
		return false
	}

	if s.EndDate != nil && now.After(*s.EndDate) {
		return false
	}

	return true
}

// IsLive checks if deal of day is currently live
func (d *DealOfDay) IsLive() bool {
	if !d.IsActive {
		return false
	}

	now := time.Now()
	return now.After(d.StartTime) && now.Before(d.EndTime) && d.Stock > d.SoldCount
}

// TimeRemaining returns the time remaining for the deal
func (d *DealOfDay) TimeRemaining() time.Duration {
	return time.Until(d.EndTime)
}

// IsLive checks if flash sale is currently live
func (f *FlashSale) IsLive() bool {
	if !f.IsActive {
		return false
	}

	now := time.Now()
	return now.After(f.StartTime) && now.Before(f.EndTime)
}

// TimeRemaining returns the time remaining for the flash sale
func (f *FlashSale) TimeRemaining() time.Duration {
	return time.Until(f.EndTime)
}

// IsLive checks if showcase is currently active
func (s *Showcase360) IsLive() bool {
	if !s.IsActive {
		return false
	}

	now := time.Now()

	if s.StartTime != nil && now.Before(*s.StartTime) {
		return false
	}

	if s.EndTime != nil && now.After(*s.EndTime) {
		return false
	}

	return true
}

// IsLive checks if bundle deal is currently active
func (b *BundleDeal) IsLive() bool {
	if !b.IsActive {
		return false
	}

	now := time.Now()

	if b.StartTime != nil && now.Before(*b.StartTime) {
		return false
	}

	if b.EndTime != nil && now.After(*b.EndTime) {
		return false
	}

	return b.Stock > b.SoldCount
}

// CalculateSavings returns the amount saved
func (b *BundleDeal) CalculateSavings() float64 {
	return b.OriginalPrice - b.BundlePrice
}

File: lib/models/homepage.dart
import 'package:thyne_jewls/models/product.dart';

enum SectionType {
  bannerCarousel('banner_carousel'),
  categories('categories'),
  dealOfDay('deal_of_day'),
  flashSale('flash_sale'),
  bestSellers('best_sellers'),
  featured('featured'),
  newArrivals('new_arrivals'),
  specialOffers('special_offers'),
  brands('brands'),
  recentlyViewed('recently_viewed'),
  upcomingEvents('upcoming_events'),
  customBanner('custom_banner'),
  showcase360('showcase_360'),
  bundleDeals('bundle_deals');

  final String value;
  const SectionType(this.value);

  static SectionType fromString(String value) {
    return SectionType.values.firstWhere(
      (e) => e.value == value,
      orElse: () => SectionType.customBanner,
    );
  }
}

class HomepageSection {
  final String id;
  final SectionType type;
  final String title;
  final String subtitle;
  final int priority;
  final bool isActive;
  final Map<String, dynamic> config;
  final DateTime? startDate;
  final DateTime? endDate;

  HomepageSection({
    required this.id,
    required this.type,
    required this.title,
    this.subtitle = '',
    required this.priority,
    required this.isActive,
    required this.config,
    this.startDate,
    this.endDate,
  });

  factory HomepageSection.fromJson(Map<String, dynamic> json) {
    return HomepageSection(
      id: json['id'] ?? '',
      type: SectionType.fromString(json['type'] ?? ''),
      title: json['title'] ?? '',
      subtitle: json['subtitle'] ?? '',
      priority: json['priority'] ?? 0,
      isActive: json['isActive'] ?? false,
      config: json['config'] ?? {},
      startDate: json['startDate'] != null
          ? DateTime.parse(json['startDate'])
          : null,
      endDate: json['endDate'] != null
          ? DateTime.parse(json['endDate'])
          : null,
    );
  }
}

class DealOfDay {
  final String id;
  final String productId;
  final double originalPrice;
  final double dealPrice;
  final int discountPercent;
  final DateTime startTime;
  final DateTime endTime;
  final int stock;
  final int soldCount;
  final bool isActive;

  DealOfDay({
    required this.id,
    required this.productId,
    required this.originalPrice,
    required this.dealPrice,
    required this.discountPercent,
    required this.startTime,
    required this.endTime,
    required this.stock,
    required this.soldCount,
    required this.isActive,
  });

  Duration get timeRemaining => endTime.difference(DateTime.now());

  bool get isLive {
    final now = DateTime.now();
    return isActive &&
           now.isAfter(startTime) &&
           now.isBefore(endTime) &&
           stock > soldCount;
  }

  int get availableStock => stock - soldCount;

  factory DealOfDay.fromJson(Map<String, dynamic> json) {
    return DealOfDay(
      id: json['id'] ?? '',
      productId: json['productId'] ?? '',
      originalPrice: (json['originalPrice'] ?? 0).toDouble(),
      dealPrice: (json['dealPrice'] ?? 0).toDouble(),
      discountPercent: json['discountPercent'] ?? 0,
      startTime: DateTime.parse(json['startTime']),
      endTime: DateTime.parse(json['endTime']),
      stock: json['stock'] ?? 0,
      soldCount: json['soldCount'] ?? 0,
      isActive: json['isActive'] ?? false,
    );
  }
}

class FlashSale {
  final String id;
  final String title;
  final String description;
  final String bannerImage;
  final List<String> productIds;
  final DateTime startTime;
  final DateTime endTime;
  final int discount;
  final bool isActive;

  FlashSale({
    required this.id,
    required this.title,
    required this.description,
    required this.bannerImage,
    required this.productIds,
    required this.startTime,
    required this.endTime,
    required this.discount,
    required this.isActive,
  });

  Duration get timeRemaining => endTime.difference(DateTime.now());

  bool get isLive {
    final now = DateTime.now();
    return isActive && now.isAfter(startTime) && now.isBefore(endTime);
  }

  factory FlashSale.fromJson(Map<String, dynamic> json) {
    // Debug: Check what ID is being parsed
    final parsedId = json['id']?.toString() ?? '';
    if (parsedId.isEmpty) {
      print('⚠️ FlashSale.fromJson: ID is empty! JSON keys: ${json.keys.toList()}');
      print('⚠️ FlashSale.fromJson: Raw JSON: $json');
    } else {
      print('✅ FlashSale.fromJson: Parsed ID=$parsedId for title=${json['title']}');
    }

    return FlashSale(
      id: parsedId,
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      bannerImage: json['bannerImage'] ?? '',
      productIds: List<String>.from(json['productIds'] ?? []),
      startTime: DateTime.parse(json['startTime']),
      endTime: DateTime.parse(json['endTime']),
      discount: json['discount'] ?? 0,
      isActive: json['isActive'] ?? false,
    );
  }
}

class Brand {
  final String id;
  final String name;
  final String logo;
  final String description;
  final bool isActive;
  final int priority;

  Brand({
    required this.id,
    required this.name,
    required this.logo,
    required this.description,
    required this.isActive,
    required this.priority,
  });

  factory Brand.fromJson(Map<String, dynamic> json) {
    return Brand(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      logo: json['logo'] ?? '',
      description: json['description'] ?? '',
      isActive: json['isActive'] ?? false,
      priority: json['priority'] ?? 0,
    );
  }
}

class Showcase360 {
  final String id;
  final String productId;
  final String title;
  final String description;
  final List<String> images360;
  final String videoUrl;
  final String thumbnailUrl;
  final int priority;
  final bool isActive;
  final DateTime? startTime;
  final DateTime? endTime;

  Showcase360({
    required this.id,
    required this.productId,
    required this.title,
    required this.description,
    required this.images360,
    required this.videoUrl,
    required this.thumbnailUrl,
    required this.priority,
    required this.isActive,
    this.startTime,
    this.endTime,
  });

  bool get isLive {
    if (!isActive) return false;
    final now = DateTime.now();
    if (startTime != null && now.isBefore(startTime!)) return false;
    if (endTime != null && now.isAfter(endTime!)) return false;
    return true;
  }

  factory Showcase360.fromJson(Map<String, dynamic> json) {
    return Showcase360(
      id: json['id'] ?? '',
      productId: json['productId'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      images360: List<String>.from(json['images360'] ?? []),
      videoUrl: json['videoUrl'] ?? '',
      thumbnailUrl: json['thumbnailUrl'] ?? '',
      priority: json['priority'] ?? 0,
      isActive: json['isActive'] ?? false,
      startTime: json['startTime'] != null
          ? DateTime.parse(json['startTime'])
          : null,
      endTime: json['endTime'] != null
          ? DateTime.parse(json['endTime'])
          : null,
    );
  }
}

class BundleItem {
  final String productId;
  final int quantity;

  BundleItem({
    required this.productId,
    required this.quantity,
  });

  factory BundleItem.fromJson(Map<String, dynamic> json) {
    return BundleItem(
      productId: json['productId'] ?? '',
      quantity: json['quantity'] ?? 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'productId': productId,
      'quantity': quantity,
    };
  }
}

class BundleDeal {
  final String id;
  final String title;
  final String description;
  final String bannerImage;
  final List<BundleItem> items;
  final double originalPrice;
  final double bundlePrice;
  final int discountPercent;
  final String category;
  final int priority;
  final bool isActive;
  final DateTime? startTime;
  final DateTime? endTime;
  final int stock;
  final int soldCount;

  BundleDeal({
    required this.id,
    required this.title,
    required this.description,
    required this.bannerImage,
    required this.items,
    required this.originalPrice,
    required this.bundlePrice,
    required this.discountPercent,
    required this.category,
    required this.priority,
    required this.isActive,
    this.startTime,
    this.endTime,
    required this.stock,
    required this.soldCount,
  });

  bool get isLive {
    if (!isActive) return false;
    if (stock <= soldCount) return false;
    final now = DateTime.now();
    if (startTime != null && now.isBefore(startTime!)) return false;
    if (endTime != null && now.isAfter(endTime!)) return false;
    return true;
  }

  int get availableStock => stock - soldCount;
  double get savings => originalPrice - bundlePrice;

  factory BundleDeal.fromJson(Map<String, dynamic> json) {
    return BundleDeal(
      id: json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      bannerImage: json['bannerImage'] ?? '',
      items: (json['items'] as List?)
          ?.map((item) => BundleItem.fromJson(item))
          .toList() ?? [],
      originalPrice: (json['originalPrice'] ?? 0).toDouble(),
      bundlePrice: (json['bundlePrice'] ?? 0).toDouble(),
      discountPercent: json['discountPercent'] ?? 0,
      category: json['category'] ?? '',
      priority: json['priority'] ?? 0,
      isActive: json['isActive'] ?? false,
      startTime: json['startTime'] != null
          ? DateTime.parse(json['startTime'])
          : null,
      endTime: json['endTime'] != null
          ? DateTime.parse(json['endTime'])
          : null,
      stock: json['stock'] ?? 0,
      soldCount: json['soldCount'] ?? 0,
    );
  }
}

class SectionLayoutItem {
  final SectionType sectionType;
  final int order;
  final bool isVisible;
  final String? title; // Optional custom title override

  SectionLayoutItem({
    required this.sectionType,
    required this.order,
    required this.isVisible,
    this.title,
  });

  factory SectionLayoutItem.fromJson(Map<String, dynamic> json) {
    return SectionLayoutItem(
      sectionType: SectionType.fromString(json['sectionType'] ?? ''),
      order: json['order'] ?? 0,
      isVisible: json['isVisible'] ?? true,
      title: json['title'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'sectionType': sectionType.value,
      'order': order,
      'isVisible': isVisible,
      if (title != null) 'title': title,
    };
  }
}

class HomepageData {
  final List<SectionLayoutItem> layout;
  final List<HomepageSection> sections;
  final DealOfDay? dealOfDay;
  final List<FlashSale> activeFlashSales;
  final List<Brand> brands;
  final List<Product> recentlyViewed;
  final List<Showcase360> showcases360;
  final List<BundleDeal> bundleDeals;

  HomepageData({
    this.layout = const [],
    required this.sections,
    this.dealOfDay,
    this.activeFlashSales = const [],
    this.brands = const [],
    this.recentlyViewed = const [],
    this.showcases360 = const [],
    this.bundleDeals = const [],
  });

  factory HomepageData.fromJson(Map<String, dynamic> json) {
    return HomepageData(
      layout: (json['layout'] as List?)
          ?.map((item) => SectionLayoutItem.fromJson(item))
          .toList() ?? [],
      sections: (json['sections'] as List?)
          ?.map((section) => HomepageSection.fromJson(section))
          .toList() ?? [],
      dealOfDay: json['dealOfDay'] != null
          ? DealOfDay.fromJson(json['dealOfDay'])
          : null,
      activeFlashSales: (json['activeFlashSales'] as List?)
          ?.map((sale) => FlashSale.fromJson(sale))
          .toList() ?? [],
      brands: (json['brands'] as List?)
          ?.map((brand) => Brand.fromJson(brand))
          .toList() ?? [],
      recentlyViewed: (json['recentlyViewed'] as List?)
          ?.map((product) => Product.fromJson(product))
          .toList() ?? [],
      showcases360: (json['showcases360'] as List?)
          ?.map((showcase) => Showcase360.fromJson(showcase))
          .toList() ?? [],
      bundleDeals: (json['bundleDeals'] as List?)
          ?.map((bundle) => BundleDeal.fromJson(bundle))
          .toList() ?? [],
    );
  }
}
